### PR TITLE
Errors for CTEs with search clause

### PR DESCRIPTION
**Pull request by halilozanakgul**
Relevant PG commit:
3696a600e2292d43c00949ddf0352e4ebb487e5b

### CODE REVIEW

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -158,6 +158,9 @@ static List * ExtractInsertValuesList(Query *query, Var *partitionColumn);
 static DeferredErrorMessage * DeferErrorIfUnsupportedRouterPlannableSelectQuery(
 	Query *query);
 static DeferredErrorMessage * ErrorIfQueryHasUnroutableModifyingCTE(Query *queryTree);
+#if PG_VERSION_NUM >= PG_VERSION_14
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/planner/multi_router_planner.c:161]: ping on this, I'm fine with merging like this, mostly curious

[the remark this answers is not in the document]
Comment (halilozanakgul) [reply src/backend/distributed/planner/multi_router_planner.c:161]: TBH I couldn't really find the root of this problem. It could be easy but right now I don't know.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 2]
```diff
@@ -3756,6 +3767,32 @@ ErrorIfQueryHasUnroutableModifyingCTE(Query *queryTree)
...
+{
+	Assert(queryTree->commandType == CMD_SELECT);
+
+	CommonTableExpr *cte = NULL;
+	foreach_ptr(cte, queryTree->cteList)
+	{
+		if (cte->search_clause != NULL)
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "cannot router plan CTE with search clause",
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:3786]: minor: the error message is hard to follow

File: src/test/regress/expected/pg14.out [hunk 3]
```diff
@@ -226,5 +226,34 @@ SELECT attname || ' ' || attcompression FROM pg_attribute WHERE attrelid::regcla
...
+WITH RECURSIVE search_graph(f, t, label) AS (
+    SELECT * FROM graph0 g WHERE f = 1
+    UNION ALL
+    SELECT g.*
+        FROM graph0 g, search_graph sg
+        WHERE g.f = sg.t and  g.f = 1
+) SEARCH DEPTH FIRST BY f, t SET seq
+SELECT * FROM search_graph ORDER BY seq;
+ERROR:  recursive CTEs are not supported in distributed queries
+WITH RECURSIVE search_graph(f, t, label) AS (
```
Comment (onderkalaci) [inline src/test/regress/expected/pg14.out:249]: This query fails if the table is a graph0 is reference table:
```SQL
WITH RECURSIVE search_graph(f, t, label) AS (                                                        SELECT * FROM graph0 g WHERE f = 1                                                                                                         UNION ALL                                                                                                                                  SELECT g.*                                                                                                                                     FROM graph0 g, search_graph sg                                                                                                             WHERE g.f = sg.t and  g.f = 1                                                                                                      ) SEARCH DEPTH FIRST BY f, t SET seq                                                                                                       DELETE FROM graph0 WHERE t IN (SELECT t FROM search_graph ORDER BY seq);
DEBUG:  Creating router plan
DEBUG:  assigned task 0 to node localhost:9700
DEBUG:  opening 1 new connections to localhost:9700
DEBUG:  established connection to localhost:9700 for session 33 in 4866 microseconds
DEBUG:  opening 1 new connections to localhost:9701
DEBUG:  opening 1 new connections to localhost:9702
DEBUG:  opening 1 new connections to localhost:9703
DEBUG:  established connection to localhost:9703 for session 36 in 34957 microseconds
DEBUG:  opening 1 new connections to localhost:9704
DEBUG:  opening 1 new connections to localhost:9705
DEBUG:  opening 1 new connections to localhost:9706
ERROR:  column reference "seq" is ambiguous
CONTEXT:  while executing command on localhost:9700
Time: 49.293 ms

```

File: src/backend/distributed/planner/multi_router_planner.c [hunk 4]
```diff
@@ -3756,6 +3777,34 @@ ErrorIfQueryHasUnroutableModifyingCTE(Query *queryTree)
...
+
+/*
+ * ErrorIfQueryHasCTEWithSearchClause checks if the query contains any common table
+ * expressions with search clause and errors out if it does.
+ */
+static DeferredErrorMessage *
+ErrorIfQueryHasCTEWithSearchClause(Query *queryTree)
+{
+	CommonTableExpr *cte = NULL;
+	foreach_ptr(cte, queryTree->cteList)
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:3790]: This check does not cover CTEs that are not in the top level:
```SQL


SELECT * FROM (
WITH RECURSIVE search_graph(f, t, label) AS
  (SELECT *
   FROM graph0 g
   WHERE f = 1 
   UNION ALL SELECT g.*
   FROM graph0 g,
        search_graph sg
   WHERE g.f = sg.t
     AND g.f = 1  ) SEARCH DEPTH FIRST BY f, t
SET seq
SELECT *
FROM search_graph
ORDER BY seq) as foo;
ERROR:  column reference "seq" is ambiguous
CONTEXT:  while executing command on localhost:9701

SELECT * FROM (
WITH RECURSIVE search_graph(f, t, label) AS
  (SELECT *
   FROM graph0 g
   WHERE f = 1 
   UNION ALL SELECT g.*
   FROM graph0 g,
        search_graph sg
   WHERE g.f = sg.t
     AND g.f = 1  ) SEARCH DEPTH FIRST BY f, t
SET seq
SELECT *
FROM search_graph
ORDER BY seq) as foo;
ERROR:  column reference "seq" is ambiguous
CONTEXT:  while executing command on localhost:9701
```

I think we need to traverse the query tree. But, we should only do if `hasRecursive` is true, to keep the overhead only applicable to queries with recursive ctes

Comment (halilozanakgul) [reply src/backend/distributed/planner/multi_router_planner.c]: I think `hasRecursive` is only valid for the current `Query`. It doesn't reflect all of the subqueries, so I couldn't use it.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 5]
```diff
@@ -3756,6 +3778,57 @@ ErrorIfQueryHasUnroutableModifyingCTE(Query *queryTree)
...
+#if PG_VERSION_NUM >= PG_VERSION_14
+
+/*
+ * ErrorIfQueryHasCTEWithSearchClause checks if the query contains any common table
+ * expressions with search clause and errors out if it does.
+ */
+static DeferredErrorMessage *
+ErrorIfQueryHasCTEWithSearchClause(Query *queryTree)
+{
+	if (query_tree_walker(queryTree, ContainsSearchClauseWalker, NULL, 0))
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:3790]: I don't think recursing only with query_tree_walker is sufficient. See `ExtractQueryWalker()`, we should also recurse into `expression_tree_walker` because some subqueries might be inside expressions (e.g., subqueries in `WHERE` clause)

**Review by onderkalaci — APPROVED** [review]
One day we should probably do the recursive query traversals just once per query so that we avoid adding such new traversals every-time we need something new.
